astropy/tests/helper.py

```diff
@@ -412,11 +413,15 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__()
```

Comment:
```suggestion
        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
        # with required argument(s) are passed to this function.
        # For a class that has no instance __dict__ and no __slots__, the default state is None
        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
```
The error in cosmology/tests/test_pickle.py was raised by recursion to a function property, which has `__getstate__()` `None`.